### ISSUE TITLE
support n-api v6

### ISSUE THREAD

**Issue by sehz** [infinyon/node-bindgen#28]

**Comment by simlay** [comment]
After a bit of review of the project and some thought, it seems that adding N-API v7 to nj-sys is very trivial and adding it to `js-core` doesn't look too bad either. I figure it's best to get some clarity on what we'd like to support or not to support. A lot of projects use feature flags to add support to new backend features such as [clang-sys](https://github.com/KyleMayes/clang-sys/blob/ddd3af568ed76a2af94bcc3a9fd06af8e00fec99/Cargo.toml#L19-L32).

The [version matrix for N-API](https://nodejs.org/api/n-api.html#n_api_n_api_version_matrix) shows, not all versions of node support the latest N-API. So, if node-bindgen had things like like `NAPI_V5` and `NAPI_V6`, this would allow for compile time checks of if the rust side will actually work with that version of node. Because all N-API versions are supersets of each other (at least from what I can tell), when you'd specify `NAPI_V6` what that really means is "not N-API v7".

Anyway, I'm not sure what the thoughts are on supporting older versions of node are but I just wanted to bring it up. Thoughts?

**Comment by sehz** [comment]
Yes, I thought putting feature flag for specific version of N-API but I think it becomes very challenge to compile one version against another as I found out.

Instead, it will be better if put `dynamic version check` for individual API.   For example, `napi_threadsafe_function` is supported for NAPI version 4 which works against node version `v10.16.0`.   So if application try to run on very older version of Node (say `v8.0'), it can return error instead of crashing.   This requires we put version check for every `NAPI` call.
